### PR TITLE
Subnet support

### PR DESCRIPTION
**Pull request by bigs**
this change introduces:

- more complete subnet support
  - linking between subnets
  - configurable ip masks at the peer and subnet level
- side effects tracking and cleanup via an easily json serializable `RenderedNetwork`

### CODE REVIEW

**Comment by bigs** [comment]
this is already a pretty good usability improvement. other things i'd like to get done (perhaps in a subsequent PR to stop it getting too large?):

- thorough namespacing to avoid conflicts with anything else running on host system
  - only bridges need to sit in the global namespace, everything else can be tucked away
  - implement a fresh name generator
  - get name reuse by generating a name, moving it into a namespace
  - track names by namespace
  - helps a lot with interface length limit (16 chars)

File: netdefine.go [hunk 1]
```diff
@@ -177,126 +275,167 @@ func Create(cfg *Config) error {
...
 			if err := l.Parse(); err != nil {
-				return err
+				return nil, err
+			}
+		}
+	}
+
+	for name, net := range nets {
+		for targetNet := range net.Links {
+			if _, ok := nets[targetNet]; !ok {
```
Comment (whyrusleeping) [inline netdefine.go]: one bit of probably helpful UX:

We could check if the name they gave was a peer, and give a special error message saying you can't link a network to a peer.

File: netdefine.go [hunk 2]
```diff
@@ -23,46 +25,124 @@ func callBin(args ...string) error {
...
+	}
+	if err := r.BridgeAddPort(b, ba); err != nil {
+		return err
+	}
+	if err := r.PortSetParameter(ba, "type", "patch"); err != nil {
+		return err
+	}
+	if err := r.PortSetOption(ba, "peer", ab); err != nil {
+		return err
+	}
```
Comment (whyrusleeping) [inline netdefine.go]: looks like we arent setting link options on the patch yet, right?

**Comment by whyrusleeping** [comment]
This is really awesome! I'm excited to take it for a spin :)

One thing to check, is that it looks like the netdef binary (in the subpackage) will no longer build. We should keep that working, its pretty useful

**Comment by bigs** [comment]
cool, will get that patched up sunday!

**Comment by bigs** [comment]
and ya i need to decide on a convention for patch link options. got some
ideas tho!

File: netdef/main.go [hunk 3]
```diff
@@ -23,16 +30,28 @@ func main() {
...
 			}
+			fi.Close()
 
-			if err := netdef.Create(&cfg); err != nil {
+			r, err := cfg.Create()
+			if err != nil {
+				return err
+			}
+
+			fi, err = os.Open(c.String("output"))
```
Comment (whyrusleeping) [inline netdef/main.go]: I would maybe encapsulate this logic in a separate function, that way we don't have to reuse the variable names

File: netdefine.go [hunk 4]
```diff
@@ -87,27 +381,45 @@ func (n *Network) GetNextIp() (string, error) {
 	ipn.Add(ipn, big.NewInt(n.nextIp))
 
 	b := ipn.Bytes()
+	subnetMask := net.IPMask(net.ParseIP(mask))
```
Comment (whyrusleeping) [inline netdefine.go:384]: whats this doing?

**Review by whyrusleeping — APPROVED** [review]
This looks pretty good to me! I'll have to give it a spin :)

File: netdef/main.go (hunk 3 above)
Comment (bigs) [reply netdef/main.go]: can do

File: netdefine.go (hunk 4 above)
Comment (bigs) [reply netdefine.go:384]: this sets the subnet mask for the interface, preferring one defined on a peer, then on its network, then the one encoded in the address space itself. essentially, if you bind to /24 you only have visibility within that /24. this seeks to address that. not sure if it's the best approach

**Pull request merged**